src/analysis/analytic/isolated_zeros.lean

```diff
@@ -0,0 +1,286 @@
+/-
+Copyright (c) 2022 Vincent Beffara. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Vincent Beffara
+-/
+import analysis.calculus.dslope
+import analysis.calculus.fderiv_analytic
+import analysis.complex.basic
+
+/-!
+# Principle of isolated zeros
+
+This file proves the fact that the zeros of a non-constant analytic function of one variable are
+isolated. It also introduces a little bit of API in the `formal_multilinear_series` namespace that
+is specific to this setup.
+
+## Main definitions
+
+When `p` is a formal multilinear series from `𝕜` to `E`:
+* `p.coef n` is the `n`th coefficient of `p` seen as a power series in `𝕜`, which is equal to `p n
+  (λ i : fin n, 1)`;
+* `p.order` (taking values in `with_top ℕ`) is the index of the first non-zero coefficient of the
+  series, or `⊤` it `p n` is identically `0`. This is the order of the zero of an analytic function
+  `f` at a point if `p` is the Taylor series of `f` at that point;
+* `p.fslope` is the series obtained by dropping the term of order `0` and dividing by the parameter,
+  corresponding to the Taylor series of `dslope f` if `p` is the Taylor series of `f`.
+
+## Main results
+
+* `has_fpower_series_at_iff` states that `has_fpower_series_at f p z₀` is equivalent to `f` being
+  locally the sum of `p`, in the sense that `∀ᶠ z in 𝓝 0, has_sum (λ n, p n (λ _, z)) (f (z₀ + z))`
+  (this version is easier to work with in some setups).
+* `analytic_at.eventually_eq_zero_or_eventually_ne_zero` is the main statement that if a function is
+  analytic at `z₀`, then either it is identically zero in a neighborhood of `z₀`, or it does not
+  vanish in a punctured neighborhood of `z₀`.
+-/
+
+open filter function nat
+open_locale topological_space big_operators
+
+variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+  {E : Type*} [normed_group E] [normed_space 𝕜 E] {s : E}
+  {p : formal_multilinear_series 𝕜 𝕜 E} {f : 𝕜 → E}
+  {n : ℕ} {z z₀ : 𝕜} {y : fin n → 𝕜}
+
+namespace formal_multilinear_series
+
+/-- The `n`th coefficient of `p` when seen as a power series. -/
+def coef (p : formal_multilinear_series 𝕜 𝕜 E) (n : ℕ) : E := p n 1
+
+@[simp] lemma apply_eq_prod_smul_coef : p n y = (∏ i, y i) • p.coef n :=
+begin
+  convert (p n).to_multilinear_map.map_smul_univ y 1,
+  funext; simp only [pi.one_apply, algebra.id.smul_eq_mul, mul_one],
+end
+
+@[simp] lemma apply_eq_pow_smul_coef : p n (λ _, z) = z ^ n • coef p n :=
+by simp
+
+@[simp] lemma norm_eq_norm_coef : ∥p n∥ = ∥coef p n∥ :=
+begin
+  apply le_antisymm,
+  { refine (p n).op_norm_le_bound (norm_nonneg (coef p n)) (λ y, _); simp [norm_smul, mul_comm] },
+  { apply le_of_le_of_eq ((p n).le_op_norm 1); simp }
+end
+
+/-- The index of the first non-zero coefficient in `p`. -/
+noncomputable def order (p : formal_multilinear_series 𝕜 𝕜 E) : with_top ℕ :=
+by classical; exact dite (∀ n, p.coef n = 0) (λ _, ⊤) (λ h, nat.find (not_forall.mp h))
```

Comment:
I think that we should have `formal_multilinear_series.order` for any multilinear series. Then you can prove that the sum of a series is its first term plus `O(...)`. BTW, you can use `supr` or `infi` to define it without `dite`.

Reply:
Yes, indeed it makes sense to generalize the definition to the general case, although I was a bit afraid of matching `p.coef n = 0` with `p n = 0`.

Saying `order := Inf { n | p n = 0 }` means that the order of the zero series would be `0` rather than `\top`, but in retrospect I am leaning more and more towards this option.

Reply:
I agree that it would make sense to have this for a general formal multilinear series. And having `order 0 = 0` is probably easier to use, indeed. Would you agree to make this change (and move this definition to another file)?